src/Microsoft.ML.Transforms/Text/NgramHashingTransformer.cs

```diff
@@ -1028,12 +970,193 @@ public VBuffer<ReadOnlyMemory<char>>[] SlotNamesMetadata(out ColumnType[] types)
+        {
+        }
+
+        /// <summary>
+        /// Produces a bag of counts of hashed ngrams in <paramref name="inputColumns"/>
+        /// and outputs ngram vector as <paramref name="outputColumn"/>
+        ///
+        /// <see cref="NgramHashingEstimator"/> is different from <see cref="WordHashBagEstimator"/> in a way that <see cref="NgramHashingEstimator"/>
+        /// takes tokenized text as input while <see cref="WordHashBagEstimator"/> tokenizes text internally.
+        /// </summary>
+        /// <param name="env">The environment.</param>
+        /// <param name="inputColumns">The columns containing text to compute bag of word vector.</param>
+        /// <param name="outputColumn">The column containing output tokens.</param>
+        /// <param name="hashBits">Number of bits to hash into. Must be between 1 and 30, inclusive.</param>
+        /// <param name="ngramLength">Ngram length.</param>
+        /// <param name="skipLength">Maximum number of tokens to skip when constructing an ngram.</param>
+        /// <param name="allLengths">Whether to include all ngram lengths up to <paramref name="ngramLength"/> or only <paramref name="ngramLength"/>.</param>
+        /// <param name="seed">Hashing seed.</param>
+        /// <param name="ordered">Whether the position of each source column should be included in the hash (when there are multiple source columns).</param>
+        /// <param name="invertHash">Limit the number of keys used to generate the slot name to this many. 0 means no invert hashing, -1 means no limit.</param>
+        public NgramHashingEstimator(IHostEnvironment env,
+            string[] inputColumns,
+            string outputColumn,
+            int hashBits = 16,
+            int ngramLength = 2,
+            int skipLength = 0,
+            bool allLengths = true,
+            uint seed = 314489979,
+            bool ordered = true,
+            int invertHash = 0)
+            : this(env, new[] { (inputColumns, outputColumn) }, hashBits, ngramLength, skipLength, allLengths, seed, ordered, invertHash)
+        {
+        }
+
+        /// <summary>
+        /// Produces a bag of counts of hashed ngrams in <paramref name="columns.inputs"/>
+        /// and outputs ngram vector for each output in <paramref name="columns.output"/>
+        ///
+        /// <see cref="NgramHashingEstimator"/> is different from <see cref="WordHashBagEstimator"/> in a way that <see cref="NgramHashingEstimator"/>
+        /// takes tokenized text as input while <see cref="WordHashBagEstimator"/> tokenizes text internally.
+        /// </summary>
+        /// <param name="env">The environment.</param>
+        /// <param name="columns">Pairs of columns to compute bag of word vector.</param>
+        /// <param name="hashBits">Number of bits to hash into. Must be between 1 and 30, inclusive.</param>
```

Comment:
Technically you can have many input columns and one output column. I am not sure that pairs is the right term. Something like mapping might work. Otherwise you could just explain what it does using something like "Defines the input and output columns for the transformation."

Reply:
Are you talking about hashBits or columns?

---
In reply to: [238801145](https://github.com/dotnet/machinelearning/pull/1811#discussion_r238801145) [](ancestors = 238801145)

Reply:
columns

---
In reply to: [238839036](https://github.com/dotnet/machinelearning/pull/1811#discussion_r238839036) [](ancestors = 238839036,238801145)